source/routing/next_stop_time.cpp

```diff
@@ -61,31 +61,42 @@ bool NextStopTimeData::Arrival::is_valid(const type::StopTime& st) const {
 
 template <typename Getter>
 void NextStopTimeData::TimesStopTimes<Getter>::init(const JourneyPattern& jp, const JourneyPatternPoint& jpp) {
+    using StopTimePair = std::pair<const type::StopTime*, const type::StopTime*>;
     // collect the stop times at the given jpp
     const auto jpp_order = jpp.order;
-    stop_times.reserve(jp.discrete_vjs.size());
+    std::vector<StopTimePair> stop_times_and_earliest_stop_time;
+    stop_times_and_earliest_stop_time.reserve(jp.discrete_vjs.size());
     for (const auto& vj : jp.discrete_vjs) {
-        const auto& st = get_corresponding_stop_time(*vj, jpp_order);
+        const type::StopTime& st = get_corresponding_stop_time(*vj, jpp_order);
         if (!getter.is_valid(st)) {
             continue;
         }
-        stop_times.push_back(&st);
+        const type::StopTime& earliest_stop_time = navitia::earliest_stop_time(st.vehicle_journey->stop_time_list);
```

Comment:
This is a bigger work, but it looks like we should (and probably already do) guarantee that stop_time_list is sorted (both in time and order).

Reply:
Yep, it would make more sense to have `vj.stop_time_list ` sorted. Maybe it already is ? I did not check

Reply:
I tried to have a look, and it's hard to have a guarantee, probably also the boarding/alighting times can mess things up.
But sharing the work upfront would certainly help... Although it's a bigger job.